### ISSUE TITLE
override404 does not send the PageNotFoundException message to the controller

### ISSUE THREAD

**Issue by crustamet** [codeigniter4/CodeIgniter4#8485]
### PHP Version

8.1

### CodeIgniter4 Version

4.4.5

### CodeIgniter4 Installation Method

Manual (zip or tar.gz)

### Which operating systems have you tested for this bug?

Windows

### Which server did you use?

apache

### What happened?

Having this for example in the Config/Routing.php
```php
public ?string $override404 = 'App\Controllers\Errors\Errors::index';
```

**CI_ENVIRONMENT** or **ENVIRONMENT** must be set to **Development**
before it looked something like this

![image](https://github.com/codeigniter4/CodeIgniter4/assets/36475250/ce710255-c19b-4dce-9b8f-69e096db0c5d)
having the message into the error page

and in Production mode the message is hidden

but i want in development mode to see the error message there is no way to make this happen besides a closure
i know it can be done with a closure but isn't there any other way to do this, if doing so then what is the point of this.

`public ?string $override404` variable it is no use for me now if using closures ?

i think it should pass the message in development mode to the controller that is written in the **$override404** variable in the config....

### Steps to Reproduce

create a Error controller

change `app/Config/Routing.php` `$override404` variable to the new error controller

### Expected Output

in development mode you should see the error message with the error page from the error controller as a parameter
in production mode you should see the error page from the error controller without the message

### Anything else?

i have tried **public ?string $override404 = App\Controllers\Errors\Errors::index/$1**

maybe you should pass an array to this variable as it states in the documentation of the routes
```php
$routes->get('messages/(:segment)', 'Support\Messages::index/$1',	['as' => 'messages']);
```

**Comment by kenjis** [comment]
I don't understand what the message is.
If you show an error message only in development mode, you need to implement it in your error controller.

I don't understand the difference between closure and controller for`$override404`.
Why you can do what you want with a closure, and can't do with a controller?

If you want to get the URI path in a controller:
```php
$uri = $this->request->getPath();
```

**Comment by kenjis** [comment]
Is your CI version 4.4.5? or really 4.5?

**Comment by crustamet** [comment]
just checked, it is **4.4.5**, this is my bad but is the same problem

and to answer your other question,
before i had this 

	if(ENVIRONMENT == 'development')
	{
		$routes->set404Override(function($message) {
			$errorController = new Errors();

			return $errorController->index($message);
		});
	}
	else
	{
		$routes->set404Override(function() {
			$errorController = new Errors();

			return $errorController->index();
		});
	}
	
And there has the same method called but with the parameter and without
but by doing this what is the point now to the $override404 in the Config/Routing.php

The only thing is that is not supported is it the variable is not passed to the Controller by using this new method **$override404** in the Config/Routing.php

I want the error message in the Error controller with the **Config/Routing.php** method

I am going to test **4.5** also

**Comment by crustamet** [comment]
I can confirm that this is working the same on **4.5**

**Comment by kenjis** [comment]
Oh, I didn't know only the Closure controller can get the PageNotFoundException message.
https://github.com/codeigniter4/CodeIgniter4/blob/65c6692ef8700236f652d96e852fdd2fb436f45d/system/CodeIgniter.php#L957-L971

Your code seems to work with 4.4.5 as before. Why do you need to set in `Config/Routing.php`?

**Comment by crustamet** [comment]
isn't that the purpose of the `Config/Routing.php` to set up stuff like this in there instead on the routes with a closure.

I dont want to use a closure because i have limited access to the request and stuff and gain access with only extra code using services to load the stuff.

There are some reasons by wanting to switch to this new Config/Routing.php file, first of all it is more organized with this new file.

im sure this can be fixed to pass the message only in dev mode using the **$override404** and stop using closures. This is the dream... this should be handled by codeigniter system as it is working without overriding the 404. that message is the hole key.

The question remains, is this a bug or not ? From my point of view it is.

![image](https://github.com/codeigniter4/CodeIgniter4/assets/36475250/ecf5592e-38ed-45f8-ac19-dbd2a941f9e6)
in this piece of code there is an documentation that states 
**It can be either a closure or the controller/method name exactly like a route is defined: Users::index**

so my Route is this : **App\Errors::show404/$1**
the thing here is only works on routes without arguments

This route is returning an error that the system wants to call `App\Errors::show404/$1()` and this does not exist... so im sure this is a bug, maybe you guys don't see it as a bug, but i surely do

**Comment by kenjis** [comment]
I think the comment is misleading. It supports only `Class::method`. Parameters like `/$1` are not supported.

As I understand it, your request is that since `$message` (the PageNotFoundException message) is passed in a 404 closure, you want `$message` to be passed in a 404 controller as well.

That is an enhancement, not a bug.

I understand why you don't want to use closures.

**Comment by crustamet** [comment]
Yes, Thanks, very hard to explain the issue, but you get it now 👍 .

Pretty big enhancement if you ask me... as you must run a controller with params
 `$returned  = $this->runController($controller, $params);`
or something to make this work i think...

**Comment by kenjis** [comment]
@crustamet I sent PR #8491

**Comment by crustamet** [comment]
Yeap this seems to be working for me
Thanks, so simple yet so effective.

`$returned = $controller->{$this->method}($e->getMessage());`

Just wow, never been thinking about this, i always think before the times when there was no possible way to run a controller just by calling it dynamically like that.

**Comment by crustamet** [comment]
just a minor adjustment
as i know you did not test this in both environments :)))

`$returned = $controller->{$this->method}((ENVIRONMENT !== 'production' || ! $this->isWeb()) ? $e->getMessage() : null);`

**Comment by kenjis** [comment]
I sent the PR for consistency. The 404 closure always receive the message. So the controller should do the same.

And I think the code like `ENVIRONMENT !== 'production'` is not good.
We cannot change the constant during testing, we cannot write test code for production environment.

Also, it is better that devs can control fully as they like. If you don't want to show the message only in production,
you can write code like that in your 404 controller or closure.

**Comment by kenjis** [comment]
#8491 is not yet merged. If nobody approve it, it will not be merged.
If you believe the PR should be merged, approve it.
See https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews

**Comment by kenjis** [comment]
Closed by #8491
